src/main/java/com/simibubi/create/content/logistics/trains/TrackMaterial.java

```diff
@@ -0,0 +1,146 @@
+
+public class TrackMaterial {
+	public static final Map<ResourceLocation, TrackMaterial> ALL = new HashMap<>();
+
+	public static final TrackMaterial ANDESITE = make(Create.asResource("andesite"))
+			.lang("Andesite")
+			.block(NonNullSupplier.lazy(() -> AllBlocks.TRACK))
+			.particle(Create.asResource("block/palettes/stone_types/polished/andesite_cut_polished"))
+			.defaultModels()
+			.build();
+
+	public final ResourceLocation id;
+	public final String langName;
+	public final NonNullSupplier<NonNullSupplier<? extends TrackBlock>> trackBlock;
+	public final Ingredient sleeperIngredient;
+	public final Ingredient railsIngredient;
+	public final ResourceLocation particle;
+	public final TrackType trackType;
+
+	@OnlyIn(Dist.CLIENT)
+	protected TrackModelHolder modelHolder;
+
+	@OnlyIn(Dist.CLIENT)
+	public TrackModelHolder getModelHolder() {
+		return modelHolder;
+	}
+
+	public TrackMaterial(ResourceLocation id, String langName, NonNullSupplier<NonNullSupplier<? extends TrackBlock>> trackBlock,
+						 ResourceLocation particle, Ingredient sleeperIngredient, Ingredient railsIngredient,
+						 TrackType trackType, Supplier<Supplier<TrackModelHolder>> modelHolder) {
+		this.id = id;
+		this.langName = langName;
+		this.trackBlock = trackBlock;
+		this.sleeperIngredient = sleeperIngredient;
+		this.railsIngredient = railsIngredient;
+		this.particle = particle;
+		this.trackType = trackType;
+		DistExecutor.unsafeRunWhenOn(Dist.CLIENT, () -> () -> this.modelHolder = modelHolder.get().get());
+		ALL.put(this.id, this);
+	}
+
+	public NonNullSupplier<? extends TrackBlock> getTrackBlock() {
+		return this.trackBlock.get();
+	}
+
+	public TrackBlock createBlock(BlockBehaviour.Properties properties) {
+		return this.trackType.factory.create(properties, this);
+	}
+
+	public boolean isCustom(String modId) {
+		return this.id.getNamespace().equals(modId);
+	}
+
+	public static TrackMaterial[] allCustom(String modid) {
+		return ALL.values().stream().filter(tm -> tm.isCustom(modid)).toArray(TrackMaterial[]::new);
+	}
+
+	public static List<NonNullSupplier<? extends TrackBlock>> allCustomBlocks(String modid) {
+		List<NonNullSupplier<? extends TrackBlock>> list = new ArrayList<>();
+		for (TrackMaterial material : allCustom(modid)) {
+			list.add(material.getTrackBlock());
+		}
+		return list;
+	}
+
+	public static List<NonNullSupplier<? extends TrackBlock>> allBlocks() {
+		List<NonNullSupplier<? extends TrackBlock>> list = new ArrayList<>();
+		for (TrackMaterial material : ALL.values()) {
+			list.add(material.getTrackBlock());
+		}
+		return list;
+	}
+
+	public String resourceName() {
+		return this.id.getPath();
+	}
+
+	public static TrackMaterial deserialize(String serializedName) {
+		ResourceLocation id = ResourceLocation.tryParse(serializedName);
+		if (id == null) {
+			Create.LOGGER.error("Failed to parse serialized track material: "+serializedName);
+			return ANDESITE;
```

Comment:
is andesite always valid, or should this be nullable?

Reply:
track materials should never be null